cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ConstraintCheckerService.java

```diff
@@ -0,0 +1,335 @@
+  }
+
+  private class ConstraintCheckerThread extends Thread {
+    private final RetryStrategy scheduleStrategy;
+    private final int partition;
+    private final Deque<Job> readyJobs = new ArrayDeque<>();
+    private Job lastConsumed;
+    private int failureCount;
+
+    ConstraintCheckerThread(int partition) {
+      super(String.format("ConstraintCheckerService-%s", "id"));
+      // TODO: [CDAP-11370] Need to be configured in cdap-default.xml. Retry with delay ranging from 0.1s to 30s
+      scheduleStrategy =
+        co.cask.cdap.common.service.RetryStrategies.exponentialDelay(100, 30000, TimeUnit.MILLISECONDS);
+      this.partition = partition;
+    }
+
+
+    @Override
+    public void run() {
+      while (isRunning()) {
+        try {
+          long sleepTime = checkJobQueue();
+          // Don't sleep if sleepTime returned is 0
+          if (sleepTime > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepTime);
+          }
+        } catch (InterruptedException e) {
+          // sleep is interrupted, just exit without doing anything
+        }
+      }
+    }
+
+    /**
+     * Check jobs in job queue for constraint satisfaction.
+     *
+     * @return sleep time in milliseconds before next fetch
+     */
+    private long checkJobQueue() {
+      boolean emptyFetch = false;
+      try {
+        emptyFetch = Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            return checkJobConstraints(context);
+          }
+        });
+        failureCount = 0;
+      } catch (Exception e) {
+        LOG.warn("Failed to check Job constraints. Will retry in next run", e);
+        failureCount++;
+      }
+
+      try {
+        Transactions.execute(transactional, new TxCallable<Boolean>() {
+          @Override
+          public Boolean call(DatasetContext context) throws Exception {
+            // run any ready jobs
+            return runReadyJobs(context);
+          }
+        });
+      } catch (Exception e) {
+        LOG.warn("Failed to launch programs. Will retry in next run", e);
+        failureCount++;
+      }
+
+      // If there is any failure, delay the next fetch based on the strategy
+      if (failureCount > 0) {
+        // Exponential strategy doesn't use the time component, so doesn't matter what we passed in as startTime
+        return scheduleStrategy.nextRetry(failureCount, 0);
+      }
+
+      // Sleep for 2 seconds if there's no jobs in the queue
+      return emptyFetch ? 2000L : 0L;
+    }
+
+    private boolean checkJobConstraints(DatasetContext context) throws Exception {
+      boolean emptyScan = true;
+
+      JobQueueDataset jobQueue = getJobQueue(context);
+      CloseableIterator<Job> jobQueueIter;
+      if (lastConsumed == null) {
+        jobQueueIter = jobQueue.getJobs(partition);
+      } else {
+        jobQueueIter = jobQueue.getJobs(partition, lastConsumed);
+      }
+
```

Comment:
perhaps here you can remember lastConsumed and reset it in case of a tx conflict